### PR TITLE
Use O_APPEND and FILE_APPEND_DATA to allow for atomic file appends

### LINKED ISSUE #53432: Allow for atomic appends to end of file
## Discussion

While working on a new API called `File.OpenHandle` I've realized that `FileStream` currently does not support handles that have been opened for appending (`FILE_APPEND_DATA` on Windows and `O_APPEND` on Unix).

Moreover, we don't use these flags for `FileMode.Append` when we are opening a file from a path. Instead, we mimic the append by setting the file offset to the end of the file:

https://github.com/dotnet/runtime/blob/33b0f242a43d56a37cedc19f95fb0a9c1f23520e/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/WindowsFileStreamStrategy.cs#L240-L243

https://github.com/dotnet/runtime/blob/33b0f242a43d56a37cedc19f95fb0a9c1f23520e/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.Unix.cs#L86-L90

https://github.com/dotnet/runtime/blob/33b0f242a43d56a37cedc19f95fb0a9c1f23520e/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.Windows.cs#L85-L88

I did some quick web search and it turns out that it was possible with .NET Framework: https://stackoverflow.com/questions/1862309/how-can-i-do-an-atomic-write-append-in-c-or-how-do-i-get-files-opened-with-the

And some of our users (@nblumhardt) are missing it in .NET Core: https://nblumhardt.com/2016/08/atomic-shared-log-file-writes/

@stephentoub @ericstj @JeremyKuhne is there any chance that you remember why we have decided to implement it in such a way?

cc @Jozkee @carlossanlop

### PR DESCRIPTION
**Pull request by adamsitnik**
So far, our `FileMode.Append` implementation was always seeking to the end of file, saving it as the current position and later using this position for writes. It was assuming that nobody else can be writing to the same file (which does not need to always be the true). 

Moreover, none of the `FileStream` ctors that accept handle accepts a `FileMode`. This information was missing, and `FileStream` created from handle opened with `O_APPEND` or `FILE_APPEND_DATA` could not recognize that it's supposed to append to the end of file.

By using `O_APPEND` and `FILE_APPEND_DATA` we can force the OS to perform atomic writes to the end of file for us.

This can be combined with disabled buffering (`bufferSize == 0`) and allow for atomic appends to file, which could be very useful for some loggers like https://github.com/serilog/serilog-sinks-file/issues/221

fixes #53432

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs [hunk 1]
```diff
@@ -86,7 +86,10 @@ internal static unsafe int WriteAtOffset(SafeFileHandle handle, ReadOnlySpan<byt
                 // The Windows implementation uses WriteFile, which ignores the offset if the handle
                 // isn't seekable.  We do the same manually with PWrite vs Write, in order to enable
                 // the function to be used by FileStream for all the same situations.
-                int result = handle.CanSeek ?
+
+                // POSIX requires that pwrite should respect provided offset even for handles opened with O_APPEND.
+                // But Linux and BSD don't do that, moreover their behaviour is different. So we always use write for O_APPEND.
+                int result = handle.CanSeek && !handle.IsAppend ?
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs:92]: Doesn't this add yet another in some cases (in particular for a `new SafeFileHandle(fd, ...)`-constructed handle?  Is that really worth it?

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+            WriteAsyncCore(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();
+
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken) =>
+#pragma warning disable CA2012 // The analyzer doesn't know the internal AsValueTask is safe.
+            WriteAsyncCore(source, cancellationToken).AsValueTask();
+#pragma warning restore CA2012
+
+        private ValueTask<int> WriteAsyncCore(ReadOnlyMemory<byte> source, CancellationToken cancellationToken)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs:183]: Does RandomAccess.WriteAtOffsetAsync guarantee to write out everything it was given, or might it write less?  If the latter, I'm realizing now that both this and the UnixFileStreamStrategy are potentially broken, in that their WriteAsync methods could then end up doing only a partial write even though the stream contract guarantees everything will be written.

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+        public override bool CanRead => false; // this is how FileMode.Append works
+
+        public override bool CanWrite => !_fileHandle.IsClosed;
+
+        public override long Length => RandomAccess.GetFileLength(_fileHandle);
+
+        public override long Position
+        {
+            get => Length;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs:43]: I don't understand this logic.  On both .NET Framework 4.8 and .NET 5, this:
```C#
using System;
using System.IO;

class Program
{
    static void Main()
    {
        const string Path = @"tmp.txt";
        File.Delete(Path);

        using (var fs = new FileStream(Path, FileMode.Append))
        {
            fs.Write(new byte[] { 1, 2, 3, 4, 5 }, 0, 5);
            fs.Position = 1;
            Console.WriteLine(fs.Position);
            Console.WriteLine(fs.Length);
        }
    }
}
```
prints:
```
1
5
```
With the logic you have here, won't this print:
```
5
5
```
?

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                FileStreamHelpers.ThrowInvalidArgument(_fileHandle);
+            }
+
+            // Prevent users from overwriting data in a file that was opened in append mode.
+            if (newLength < length)
+            {
+                throw new IOException(SR.IO_SeekAppendOverwrite);
+            }
+
+            FileStreamHelpers.SetFileLength(_fileHandle, newLength);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs:119]: I also don't understand this logic.  This program:
```C#
using System;
using System.IO;
using System.Text;

class Program
{
    static void Main()
    {
        const string Path = @"tmp.txt";
        File.Delete(Path);

        using (var fs = new FileStream(Path, FileMode.Append))
        {
            fs.Write(Encoding.ASCII.GetBytes("hello"));
            fs.Position = 1;
            fs.Write(Encoding.ASCII.GetBytes("abc"));
        }

        Console.WriteLine(File.ReadAllText(Path));
    }
}
```
today prints:
```
habco
```
highlighting that a) you are allowed to seek backwards (just not before the position when the file was opened) and b) it's not losing all data after the seek point.

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        internal override void Lock(long position, long length) => FileStreamHelpers.Lock(_fileHandle, CanWrite, position, length);
+
+        internal override void Unlock(long position, long length) => FileStreamHelpers.Unlock(_fileHandle, position, length);
+
+        public override void SetLength(long value)
+        {
+            if (value < Length)
+            {
+                throw new IOException(SR.IO_SetLengthAppendTruncate);
+            }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs:133]: Similarly, this seems wrong.  Today this:
```C#
using System;
using System.IO;
using System.Text;

class Program
{
    static void Main()
    {
        const string Path = @"tmp.txt";
        File.Delete(Path);

        using (var fs = new FileStream(Path, FileMode.Append))
        {
            fs.Write(Encoding.ASCII.GetBytes("hello"));
            fs.SetLength(1);
        }

        Console.WriteLine(File.ReadAllText(Path));
    }
}
```
prints
```
h
```
but it seems like this code will end up throwing.

**Review by stephentoub — CHANGES_REQUESTED** [review]

**Comment by adamsitnik** [comment]
I am going to close the PR right now and re-open it when I am back from vacations.

**Comment by adamsitnik** [comment]
The official [doc](https://docs.microsoft.com/en-us/dotnet/api/system.io.filemode?view=net-5.0) for `FileMode.Append` says:

> Trying to seek to a position before the end of the file throws an IOException exception

So setting `Postion` to anything that is less that `Length` should throw. I believe that existing implementation was incomplete.

The position was captured when file was being opened:

https://github.com/dotnet/runtime/blob/b75e55beb1628df8f704a3014b9acd1b98d4f445/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.Windows.cs#L48-L55

We were trying to prevent from truncating the file:
https://github.com/dotnet/runtime/blob/b75e55beb1628df8f704a3014b9acd1b98d4f445/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.Windows.cs#L253-L254

https://github.com/dotnet/runtime/blob/b75e55beb1628df8f704a3014b9acd1b98d4f445/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.Windows.cs#L415-L421

The problem is that we were never updating the `_appendStart` value which should be happening after each write.

The way I understand `FileMode.Append`: always append to the end of file, don't allow for truncation or for overwriting the data.

@JeremyKuhne @stephentoub is my understanding correct?

**Comment by stephentoub** [comment]
I don't agree with the interpretation.  I read the docs as being about seeking to earlier than the position when the file was opened,  which is exactly what the implementation supports. Changing it is a breaking change.

**Comment by adamsitnik** [comment]
> also in required permissions.

Could you please explain what do you mean by that?

**Comment by snakefoot** [comment]
@adamsitnik I can understand the wish to make the FileStream-contract predictable, when applying FILE_APPEND_DATA / O_APPEND where Operating System always appends to end (and ignore any file-seeks or change of file-position).

When the user has opened a file with `FileMode.Append` then I think the user will accept that even if the doing custom file-seeks, then all writes will be appended to the end (independent of current file-position).

But it will be confusing for the user if not possible to make random reads. For me `FileMode.Append` means always write to the end of the file, but still allow random reads.

- When making a file-write then FileStream-Position should automatically move to the end when `FileMode.Append` (Perform the write and then seek to end)
- When checking current FileStream-Position then it will remain unchanged (even if other processes might have appended to the file).
- When making a file-read then it will always use the current FileStream-Position

**Comment by JeremyKuhne** [comment]
I think there is some unfortunate concept overlap between specific rights and "position at the end of the file". It looks as if we were trying to emulate `FILE_APPEND_DATA`'s behavior over `GENERIC_WRITE`, and not quite getting it right.

I wouldn't change the existing behavior. My opinion is clarify the docs and add a new option if we think differing functionality is important.

**Comment by stephentoub** [comment]
> Could you please explain what do you mean by that?

I mean FILE_APPEND_DATA is a specific access right:
https://docs.microsoft.com/en-us/windows/win32/fileio/file-access-rights-constants
representing "the right to append data to the file".  I don't know what the ramifications in this regard would be of suddently changing FileMode.Append to specify that for a file object that might not allow it.

**Comment by snakefoot** [comment]
In NetFramework then FILE_APPEND_DATA is applied when using the FileStream-constructor with [FileSystemRights.AppendData](https://docs.microsoft.com/en-us/dotnet/api/system.security.accesscontrol.filesystemrights)

**Comment by adamsitnik** [comment]
> I don't know what the ramifications in this regard would be of suddently changing FileMode.Append to specify that for a file object that might not allow it.

We are currently using `GENERIC_WRITE`:

https://github.com/dotnet/runtime/blob/1be8fd886022765ba31a4a1872552baeeadf1863/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs#L64-L66

https://github.com/dotnet/runtime/blob/1be8fd886022765ba31a4a1872552baeeadf1863/src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GenericOperations.cs#L11

which already contains `FILE_APPEND_DATA` amongst many other things:

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs#L594-L600

I am not sure if contains is the right word, but it's more or less:

```cs
FILE_GENERIC_WRITE  =  FILE_APPEND_DATA | FILE_WRITE_ATTRIBUTES | FILE_WRITE_DATA | FILE_WRITE_EA | STANDARD_RIGHTS_READ and so on...
```

So permissions would not be a problem

**Comment by adamsitnik** [comment]
> I wouldn't change the existing behavior. My opinion is clarify the docs and add a new option if we think differing functionality is important.

You are right. Let me prepare a proposal. I am going to convert this PR into a draft and add `NO MERGE` label until then.

**Comment by adamsitnik** [comment]
Done: https://github.com/dotnet/runtime/issues/53432#issuecomment-902478772

**Pull request closed without merging**